### PR TITLE
client: serverless mode + refactor into plugin

### LINKED ISSUE #27: client: serverless mode
Make the `--server` command-line option optional so that the client can run without connecting to a TCP server.

In order for the IPC daemon to still run, the client must have a `PeerProvider` implementation to provide with the `DaemonOffer`. `hearth-server` provides it, but `hearth-client` doesn't. We need to move `PeerProviderImpl` from the server to the core crate so that the client can use it too.

### PR DESCRIPTION
**Pull request by marceline-cramer**
Closes #27.

### CODE REVIEW

File: crates/hearth-client/src/main.rs [hunk 1]
```diff
@@ -40,9 +40,8 @@ mod window;
 #[derive(Parser, Debug)]
 pub struct Args {
     /// IP address and port of the server to connect to.
-    // TODO support DNS resolution too
```
Comment (airidaceae) [inline crates/hearth-client/src/main.rs:43]: the relics of hearth's past

File: crates/hearth-client/src/main.rs [hunk 2]
```diff
@@ -108,118 +107,136 @@ async fn async_main(args: Args, rend3_plugin: Rend3Plugin) {
...
-    builder.add_plugin(init);
     builder.add_plugin(hearth_daemon::DaemonPlugin::new());
-    let runtime = builder.run(config).await;
 
-    tokio::spawn(async move {
-        connect(network_root_rx, runtime, args.server, args.password).await;
-    });
+    if let (Some(server), password) = (args.server, args.password) {
+        builder.add_plugin(ClientPlugin { server, password });
+    }
```
Comment (airidaceae) [inline crates/hearth-client/src/main.rs:120]: add an info level log message that we are starting in serverless mode.

File: crates/hearth-client/src/main.rs [hunk 3]
```diff
@@ -108,118 +107,136 @@ async fn async_main(args: Args, rend3_plugin: Rend3Plugin) {
...
-        Err(_) => {
-            info!(
-                "Failed to parse \'{}\' to SocketAddr, attempting DNS resolution",
-                server
-            );
-            match server.to_socket_addrs() {
-                Err(err) => {
-                    error!("Failed to resolve IP: {:?}", err);
-                    return;
+pub struct ClientPlugin {
```
Comment (airidaceae) [inline crates/hearth-client/src/main.rs:128]: Add a doc comment that makes it clear that the only way for the client to connect to a server is through this plugin

**Review by airidaceae — CHANGES_REQUESTED** [review]

**Review by airidaceae — APPROVED** [review]
Looks good

**Comment by airidaceae** [comment]
going to merge this in so we can work through any conflicts on #183

**Comment by marceline-cramer** [comment]
Sounds good to me, thank you!

**Pull request merged**